pkg/postage/batchstore/reserve_test.go

```diff
@@ -0,0 +1,306 @@
+// Copyright 2021 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package batchstore_test
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"math/big"
+	"math/rand"
+	"os"
+	"testing"
+
+	"github.com/ethersphere/bee/pkg/postage"
+	"github.com/ethersphere/bee/pkg/postage/batchstore"
+	postagetest "github.com/ethersphere/bee/pkg/postage/testing"
+	"github.com/ethersphere/bee/pkg/statestore/leveldb"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+// random advance on the blockchain
+func newBlockAdvance() uint64 {
+	return uint64(rand.Intn(3) + 1)
+}
+
+// initial depth of a new batch
+func newBatchDepth(depth uint8) uint8 {
+	return depth + uint8(rand.Intn(10)) + 4
+}
+
+// the factor to increase the batch depth with
+func newDilutionFactor() int {
+	return rand.Intn(3) + 1
+}
+
+// new value on top of value based on random period and price
+func newValue(price, value *big.Int) *big.Int {
+	period := rand.Intn(100) + 1000
+	v := new(big.Int).Mul(price, big.NewInt(int64(period)))
+	return v.Add(v, value)
+}
+
+// TestBatchStoreUnreserve is testing the correct behaviour of the reserve.
+// the following assumptions are tested on each modification of the batches (top up, depth increase, price change)
+// - reserve exceeds capacity
+// - value-consistency of unreserved POs
+func TestBatchStoreUnreserveEvents(t *testing.T) {
+	// temporarily reset reserve Capacity
+	defer func(i int64) {
+		batchstore.Capacity = i
+	}(batchstore.Capacity)
+	batchstore.Capacity = batchstore.Exp2(16)
+
+	bStore, unreserved := setupBatchStore(t)
+	batches := make(map[string]*postage.Batch)
+
+	t.Run("new batches only", func(t *testing.T) {
+		// iterate starting from batchstore.DefaultDepth to maxPO
+		_, depth := batchstore.GetReserve(bStore)
+		for step := 0; depth < swarm.MaxPO; step++ {
+			cs, err := nextChainState(bStore)
+			if err != nil {
+				t.Fatal(err)
+			}
+			var b *postage.Batch
+			if b, err = createBatch(bStore, cs, depth); err != nil {
+				t.Fatal(err)
+			}
+			batches[string(b.ID)] = b
+			if depth, err = checkReserve(bStore, unreserved); err != nil {
+				t.Fatal(err)
+			}
+		}
+	})
+	t.Run("top up batches", func(t *testing.T) {
+		n := 0
+		for id := range batches {
+			b, err := bStore.Get([]byte(id))
+			if err != nil {
+				if errors.Is(storage.ErrNotFound, err) {
+					continue
+				}
+				t.Fatal(err)
+			}
+			cs, err := nextChainState(bStore)
```

Comment:
not sure why you need to be returning errors from all of these test helpers. you can just pass `t` to them, declare `t.Helper()` in the function, then `t.Fatal(err)` directly within the helper. this goes for all of this file

Reply:
ok if you prefer i can rewrite